common/src/service/ExceptionHandlingHelper.cs

```diff
@@ -116,12 +117,56 @@ public static async Task<ErrorCode> GetExceptionCodeAsync(HttpResponseMessage re
             try
             {
                 IoTHubExceptionResult responseContent = JsonConvert.DeserializeObject<IoTHubExceptionResult>(responseContentStr);
-                Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent.Message);
 
-                if (messageFields != null
-                    && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                try
                 {
-                    errorCodeValue = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                    Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent.Message);
+
+                    if (messageFields != null
+                        && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                    {
+                        errorCodeValue = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                    }
+                }
+                catch (JsonReaderException ex)
+                {
+                    if (Logging.IsEnabled)
+                        Logging.Error(null, $"Failed to deserialize error message into a dictionary: {ex}. Message body: '{responseContentStr}.'");
+
+                    // In some scenarios, the error response string is a ';' delimited string with the service-returned error code.
+                    const char errorFieldsDelimiter = ';';
+                    string[] messageFields = responseContent.Message.Split(errorFieldsDelimiter);
+
+                    if (messageFields != null)
+                    {
+                        _ = messageFields
+                            .Any(field =>
+                            {
+#if NET451 || NET472 || NETSTANDARD2_0
+                                if (field.IndexOf(CommonConstants.ErrorCode, StringComparison.OrdinalIgnoreCase) >= 0)
+#else
+                                if (field.Contains(CommonConstants.ErrorCode, StringComparison.OrdinalIgnoreCase))
+#endif
+                                {
+                                    const char errorCodeDelimiter = ':';
+                                    string[] errorCodeFields = field.Split(errorCodeDelimiter);
+
+                                    if (Enum.TryParse(errorCodeFields[1], out ErrorCode errorCode))
```

Comment:
If there is no : in the string, it could lead to IndexOutOfRangeException. Check errorCodeFields.Length > 1 before accessing errorCodeFields[1]

Reply:
This code will be executed only if there is an entry for "errorCode" in the message response, which is always a key-value pair. So the string will always have a ":". I can still add a a check though, not a problem.